src/main/java/io/dockstore/provision/GSPlugin.java

```diff
@@ -0,0 +1,271 @@
+                            bytes.flip();
+                            channel.write(bytes);
+                            bytes.clear();
+                            runningTotal = runningTotal + limit;
+                            printer.handleProgress(runningTotal, inputSize);
+                        }
+                    } catch (IOException e) {
+                        System.err.println("Could not download file. IO exception:" + e.getMessage());
+                        return false;
+                    }
+
+                }
+            }
+
+            writeTo.close();
+            return true;
+        }
+
+        public boolean uploadTo(String destPath, Path sourceFile, Optional<String> metadata) {
+            long inputSize = sourceFile.toFile().length();
+
+            String contentType = "application/octet-stream";
+            try {
+                contentType = Files.probeContentType(sourceFile);
+            } catch (IOException e) {
+                System.err.println("Could not get content type of file. Using default content type. IO exception:" + e.getMessage());
+            }
+
+            // Bug on Mac where Files.probeContentType returns null
+            if (contentType == null)
+                contentType = "application/octet-stream";
+
+            Storage gsClient = getGoogleGSClient();
+
+            String trimmedPath = destPath.replace("gs://", "");
+            List<String> splitPathList = Lists.newArrayList(trimmedPath.split("/"));
+            String bucketName = splitPathList.remove(0);
+            String blobName = String.join(File.separator, splitPathList);
+
+            BlobId blobId = BlobId.of(bucketName, blobName);
+            BlobInfo blobInfo;
+
+
+            // To test metadata uncomment the lines below
+            //Map<String,String> myMap = new HashMap<String, String>();
+            //myMap.put("goat", "bleat");
+            //myMap.put("cow", "moo");
+            //metadata = Optional.of(myMap.toString());
+
+            if (metadata.isPresent()) {
+                Gson gson = new Gson();
+                Type type = new TypeToken<Map<String, String>>() {
+                }.getType();
+
+                try {
+                    Map<String, String> map = gson.fromJson(metadata.get(), type);
+                    for (Map.Entry<String, String> entry : map.entrySet()) {
+                        System.out.println("Loading " + entry.getKey() + "->" + entry.getValue());
+                    }
+                    blobInfo = BlobInfo.newBuilder(blobId).setContentType(contentType).setMetadata(map).build();
+                } catch (com.google.gson.JsonSyntaxException ex) {
+                    System.err.println(
+                            "Could not load metadata. Metadata syntax exception. Uploading file without metadata:" + ex.getMessage());
+                    blobInfo = BlobInfo.newBuilder(blobId).setContentType(contentType).build();
```

Comment:
One could argue this both ways, but I would lean towards having the upload fail if the metadata is not correct. If you are trying to set the metadata and it fails, then I would think you would not want to proceed.

Reply:
Hmm...just left it as is

Reply:
Changed code to fail upload if metadata cannot be set